feat: devcontainer.json to work for vLLM, SGLang, and TensorRT-LLM

keivenchang opened the pull request:
#### Overview:

Convert devcontainer.json to a Jinja2 template system that generates framework-specific devcontainer configurations for vLLM, SGLang, and TensorRT-LLM.

#### Details:

- Refactor `.devcontainer/devcontainer.json` to `.devcontainer/devcontainer.json.j2` Jinja2 template
- Add `gen_devcontainer_json.py` script to generate framework-specific configurations
- Generate separate devcontainer.json files for vllm, sglang, and trtllm frameworks
- Update README.md to document the new template system and remove vLLM-only references
- Add auto-generated warning comments to prevent direct editing of generated files
- Make copyright year dynamic using current_year template variable

#### Where should the reviewer start?

- `.devcontainer/devcontainer.json.j2` - The main Jinja2 template
- `.devcontainer/gen_devcontainer_json.py` - Generation script logic
- `.devcontainer/README.md` - Updated documentation explaining the new system

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)



/coderabbit profile chill

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- New Features
  - Introduced framework-specific devcontainers for VLLM, SGLang, and TensorRT-LLM with GPU support, presets (Python/Rust extensions), and resource tuning.
  - Added persistent caches (history, pre-commit) and post-create bootstrap to streamline setup.
  - Parameterized image naming to support multiple frameworks.

- Documentation
  - Updated devcontainer README with framework-specific directories, generation workflow, and generalized build/run commands.
  - Clarified instructions for regenerating configs and cautions against manual edits of generated files.

- Chores
  - Added a generator to produce framework-specific devcontainer configurations.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

keivenchang committed:
refactor: convert devcontainer.json to Jinja2 template

- Rename .devcontainer/devcontainer.json to .devcontainer/devcontainer.json.j2
- Add template variables for framework-specific generation
- Make copyright year dynamic

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
feat: add devcontainer generation script

- Add gen_devcontainer_json.py to generate framework-specific devcontainers
- Supports vllm, sglang, and trtllm frameworks
- Automatically adds warning comments to generated files

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
feat: add generated devcontainer configs for frameworks

- Add framework-specific devcontainer.json files for vllm, sglang, trtllm
- Generated from devcontainer.json.j2 template
- Include auto-generated warning comments

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
docs: remove vLLM-only references in README

- Update build commands to use generic FRAMEWORK placeholder
- Replace hardcoded vLLM examples with framework-agnostic instructions
- Clarify that users should replace FRAMEWORK with vllm/sglang/trtllm

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
fix: update template and regenerate devcontainer configs

- Fix template variables in devcontainer.json.j2
- Regenerate framework-specific devcontainer.json files
- Ensure consistent naming and dynamic year in generated files

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
fix: add SPDX license header to generation script

- Add required SPDX-FileCopyrightText and SPDX-License-Identifier headers
- Comply with project license header requirements

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang commented on .devcontainer/gen_devcontainer_json.py:
```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env python3
+"""
+Script to generate devcontainer.json files for different frameworks using Jinja2 template.
+"""
```
Added.

biswapanda commented on .devcontainer/README.md:
```diff
@@ -89,20 +111,29 @@ You must have the following path on your host.
-./container/build.sh --dev-image dynamo:latest-vllm --framework VLLM
-# Now you have a local-dev image dynamo:latest-vllm-local-dev
+# Single command approach (recommended) - replace FRAMEWORK with VLLM/SGLANG/TRTLLM
+./container/build.sh --framework FRAMEWORK --target local-dev
```
should we provide copy-paste-able commands to end users for default framework (vLLM)

```
export FRAMEWORK=VLLM         # Note:  any of VLLM, SGLANG, TRTLLM can be used
./container/build.sh --framework $FRAMEWORK --target local-dev
```

biswapanda commented on .devcontainer/README.md:
```diff
@@ -362,9 +393,9 @@ If you see errors like "container is not running" or "An error occurred setting
-   ./container/build.sh --framework vllm
-   ./container/build.sh --dev-image dynamo:latest-vllm --framework vllm
-   # Output: dynamo:latest-vllm-local-dev
+   ./container/build.sh --framework FRAMEWORK  # Replace with vllm/sglang/trtllm
```
should we follow [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) and use `$FRAMEWORK` ? 

this becomes repetitive: ` Replace with vllm/sglang/trtllm`

keivenchang replied:
I'll make this change in the next PR.

keivenchang commented:
> `less is more` and `DRY`
> 
> what is difference across framework's dev container jsons ? is it possible to parameterize framework and have single file so there is single file to manage.

I really wished that devcontainer.json Extension (written by Microsoft) would allow this. But they say that every Dev Container needs to have this following convention, which unfortunately, is a pain and has a lot of duplications. This is why I resorted to the template system. Source: https://code.visualstudio.com/remote/advancedcontainers/connect-multiple-containers

I looked at so many different places but, it looks like this is the only way 🥺

```
📁 project-root
    📁 .git
    📁 .devcontainer
      📁 python-container
        📄 devcontainer.json
      📁 node-container
        📄 devcontainer.json
    📁 python-src
        📄 hello.py
    📁 node-src
        📄 hello.js
    📄 docker-compose.yml
```

keivenchang committed:
docs: improve devcontainer build commands with explicit variable usage

- Use 'export FRAMEWORK=VLLM' instead of inline 'FRAMEWORK' placeholders
- Change 'dynamo:latest-FRAMEWORK' to 'dynamo:latest-${FRAMEWORK,,}' for clarity
- Provide copy-paste-able commands with concrete vLLM examples
- Apply consistent format to both main instructions and troubleshooting section

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

alec-flowers commented:
> > `less is more` and `DRY`
> > what is difference across framework's dev container jsons ? is it possible to parameterize framework and have single file so there is single file to manage.
> 
> I really wished that the Dev Container Extension (written by Microsoft) would allow this. But their doc says that every `devcontainer.json` needs to have this following convention, which unfortunately, is a pain and has a lot of duplications. I experimented with a lot of different undocumented methods but they didn't seem to work.
> 
> Finally, I resorted to the template system. Source: https://code.visualstudio.com/remote/advancedcontainers/connect-multiple-containers
> 
> It looks like this is the only way for now. Maybe if enough people requests for this feature, they'll add it.
> 
> ```
> 📁 project-root
>     📁 .git
>     📁 .devcontainer
>       📁 python-container
>         📄 devcontainer.json
>       📁 node-container
>         📄 devcontainer.json
>     📁 python-src
>         📄 hello.py
>     📁 node-src
>         📄 hello.js
>     📄 docker-compose.yml
> ```

This link was quite helpful. Could you include it somewhere in the docs so people know why and how you came to this conclusion?

alec-flowers reviewed (approved):
2 comments, generally looks good. Thanks for this!

keivenchang commented on .devcontainer/README.md:
```diff
@@ -89,19 +111,36 @@ You must have the following path on your host.
 
 Follow these steps to get your NVIDIA Dynamo development environment up and running:
 
+### Step 0: Choose Your Framework
```
You're absolutely right thanks for proof-reading!

keivenchang committed:
docs: clarify devcontainer template system rationale and reorder setup steps

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>
